File: src/main/java/link/pihda/billofmaterial/dao/TransactionDao.java
/*
 * Copyright (c) 2023.  Adib S. A.
 * Copyright (c) 2023.  PT Atadopos
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */

package link.pihda.billofmaterial.dao;

import jakarta.persistence.EntityManager;
import link.pihda.billofmaterial.entity.Transaction;
import link.pihda.billofmaterial.util.JPAUtility;

import java.util.List;

public class TransactionDao {
    public void save(link.pihda.billofmaterial.entity.Transaction transaction) {
        try (EntityManager entityManager = JPAUtility.getEntityManagerFactory().createEntityManager()) {
            entityManager.getTransaction().begin();
            entityManager.persist(transaction);
            entityManager.getTransaction().commit();
        }
    }

    public List<link.pihda.billofmaterial.entity.Transaction> getAll() {
        try (EntityManager entityManager = JPAUtility.getEntityManagerFactory().createEntityManager()) {
            return entityManager.createQuery("SELECT i FROM Transaction i", Transaction.class).getResultList();
        }
    }


    public void delete(link.pihda.billofmaterial.entity.Transaction transaction) {
        try (EntityManager entityManager = JPAUtility.getEntityManagerFactory().createEntityManager()) {
            entityManager.getTransaction().begin();
            entityManager.remove(entityManager.contains(transaction) ? transaction : entityManager.merge(transaction));
            entityManager.getTransaction().commit();
        }
    }

    public void update(link.pihda.billofmaterial.entity.Transaction transaction) {
        try (EntityManager entityManager = JPAUtility.getEntityManagerFactory().createEntityManager()) {
            entityManager.getTransaction().begin();
            entityManager.merge(transaction);
            entityManager.getTransaction().commit();
        }
    }
}

File: src/main/kotlin/link/pihda/billofmaterial/dao/TransactionDao.kt
/*
 * Copyright (c) 2023.  Adib S. A.
 * Copyright (c) 2023.  PT Atadopos
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */
package link.pihda.billofmaterial.dao

import jakarta.persistence.EntityManager
import link.pihda.billofmaterial.entity.Transaction
import link.pihda.billofmaterial.util.JPAUtility

class TransactionDao {
    fun save(transaction: Transaction) {
        JPAUtility.getEntityManagerFactory().createEntityManager().use { entityManager: EntityManager ->
            entityManager.transaction.begin()
            entityManager.persist(transaction)
            entityManager.transaction.commit()
        }
    }

    fun getAll(): List<Transaction> {
        return JPAUtility.getEntityManagerFactory().createEntityManager().use { entityManager: EntityManager ->
            entityManager.createQuery("SELECT i FROM Transaction i", Transaction::class.java).resultList
        }
    }

    fun delete(transaction: Transaction) {
        JPAUtility.getEntityManagerFactory().createEntityManager().use { entityManager: EntityManager ->
            entityManager.transaction.begin()
            val managedTransaction =
                if (entityManager.contains(transaction)) transaction else entityManager.merge(transaction)
            entityManager.remove(managedTransaction)
            entityManager.transaction.commit()
        }
    }

    fun update(transaction: Transaction) {
        JPAUtility.getEntityManagerFactory().createEntityManager().use { entityManager: EntityManager ->
            entityManager.transaction.begin()
            entityManager.merge(transaction)
            entityManager.transaction.commit()
        }
    }
}

